# List Comprehensions Lab

Complete the following set of exercises to solidify your knowledge of list comprehensions.

In [8]:
import os
import numpy as np
import pandas as pd

### 1. Use a list comprehension to create and print a list of consecutive integers starting with 1 and ending with 50.

In [9]:
lst=[i for i in range(1,51)]
print(lst)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]


### 2. Use a list comprehension to create and print a list of even numbers starting with 2 and ending with 200.

In [10]:
lst2=[ i for i in range(2,201) if i %2==0]
print(lst2)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200]


### 3. Use a list comprehension to create and print a list containing all elements of the 10 x 4 Numpy array below.

In [11]:
a = np.array([[0.84062117, 0.48006452, 0.7876326 , 0.77109654],
       [0.44409793, 0.09014516, 0.81835917, 0.87645456],
       [0.7066597 , 0.09610873, 0.41247947, 0.57433389],
       [0.29960807, 0.42315023, 0.34452557, 0.4751035 ],
       [0.17003563, 0.46843998, 0.92796258, 0.69814654],
       [0.41290051, 0.19561071, 0.16284783, 0.97016248],
       [0.71725408, 0.87702738, 0.31244595, 0.76615487],
       [0.20754036, 0.57871812, 0.07214068, 0.40356048],
       [0.12149553, 0.53222417, 0.9976855 , 0.12536346],
       [0.80930099, 0.50962849, 0.94555126, 0.33364763]])

In [16]:

import numpy as np

#first atempt:
nu_array = [ i for i in a]
for b in nu_array:
    print(b)
size=(4, 10)

#why this version doesn`t work and print only the last line?
#nu_array = [ b for i in a for b in a]
#print(b)



[0.84062117 0.48006452 0.7876326  0.77109654]
[0.44409793 0.09014516 0.81835917 0.87645456]
[0.7066597  0.09610873 0.41247947 0.57433389]
[0.29960807 0.42315023 0.34452557 0.4751035 ]
[0.17003563 0.46843998 0.92796258 0.69814654]
[0.41290051 0.19561071 0.16284783 0.97016248]
[0.71725408 0.87702738 0.31244595 0.76615487]
[0.20754036 0.57871812 0.07214068 0.40356048]
[0.12149553 0.53222417 0.9976855  0.12536346]
[0.80930099 0.50962849 0.94555126 0.33364763]


### 4. Add a condition to the list comprehension above so that only values greater than or equal to 0.5 are printed.

In [20]:
#new = [v for i in b for v in i if v >= 0.5]
#print(v)

#b= a.tolist()

c=[string(v.split) for i in b for v in i if v]



TypeError: 'numpy.float64' object is not iterable

### 5. Use a list comprehension to create and print a list containing all elements of the 5 x 2 x 3 Numpy array below.

In [21]:
b = np.array([[[0.55867166, 0.06210792, 0.08147297],
        [0.82579068, 0.91512478, 0.06833034]],

       [[0.05440634, 0.65857693, 0.30296619],
        [0.06769833, 0.96031863, 0.51293743]],

       [[0.09143215, 0.71893382, 0.45850679],
        [0.58256464, 0.59005654, 0.56266457]],

       [[0.71600294, 0.87392666, 0.11434044],
        [0.8694668 , 0.65669313, 0.10708681]],

       [[0.07529684, 0.46470767, 0.47984544],
        [0.65368638, 0.14901286, 0.23760688]]])



#b.tolist() why in the exercise 3 its work and is this one dont?  its because whe have another list?


AttributeError: 'list' object has no attribute 'tolist'

### 6. Add a condition to the list comprehension above so that the last value in each subarray is printed, but only if it is less than or equal to 0.5.

In [ ]:
d= [l[-1] for ll in b for l in ll if l[-1]]
print(d)

### 7. Use a list comprehension to select and print the names of all CSV files in the */data* directory.

### 8. Use a list comprehension and the Pandas `read_csv` and `concat` methods to read all CSV files in the */data* directory and combine them into a single data frame. Display the top 10 rows of the resulting data frame.

### 9. Use a list comprehension to select and print the column numbers for columns from the data set whose median is less than 0.48.

### 10. Use a list comprehension to add a new column (20) to the data frame whose values are the values in column 19 minus 0.1. Display the top 10 rows of the resulting data frame.

### 11. Use a list comprehension to extract and print all values from the data set that are between 0.7 and 0.75.